whole-tale/girder_wholetale
# Create valid BDBags with external data

To summarize what happen in this PR, cause there are a few nuances that made it work:
1. External data in bags is unconditionally expanded, meaning no folders are recorded in `aggregates` section, only files. (see https://github.com/whole-tale/girder_wholetale/pull/518/commits/95d5b36f9d72955557b0814ae5a89322fd1eda75). This avoids situation when 1) object doesn't have a `uri` or 2) we end up with `doi:something data/folder` in bag's manifest. The latter, while neat wasn't even working with BDBag, cause it doesn't have a generic doi resolver.
2. Implementing 1. allows for a slightly easier/faster `dataSet` generation during Tale import. This is important for a roundtrip of Tales that have folders or subfolders from external datasets in `data/`. (see https://github.com/whole-tale/girder_wholetale/pull/518/commits/46a6f3073fe09e210854790e9fabbf41ba40f972)
3. I silently switched default *sha256* algorithm in bags to *sha512*, since the latter is used by Girder and will be required anyway once we start exporting files directly from assetstores. (That happened in https://github.com/whole-tale/girder_wholetale/pull/518/commits/ce194f106258355112e47ed7de2de3dd5f319fca). Note: that it breaks the profile we're using (https://raw.githubusercontent.com/fair-research/bdbag/master/profiles/bdbag-ro-profile.json) which requires md5 and sha256. Note: bdbag doesn't complain... :-) Nevertheless, we should probably use our own profile?
4. We now collect checksums for external data directly from providers and use that to create a valid manifest.txt during bag export (see https://github.com/whole-tale/girder_wholetale/pull/518/commits/9e51c1f9f7716e97e4eb3517981ca3ec97ff2409). That comes with a caveat: not every provider has that info or sometimes info is missing for some files (Dataverse and tab-files, hence the ugly workaround) There are two cases where we don't know hashes: raw HTTP/HTTPS resources and Globus. The former we handle via **downloading the data and calculating the md5sum during the Tale export**, which is potentially a long process (see `TaleExporter.verify_aggregate_checksums` in https://github.com/whole-tale/girder_wholetale/pull/518/commits/ce194f106258355112e47ed7de2de3dd5f319fca). I haven't checked how it's gonna behave for "very large data"^{TM}. Globus simply doesn't work right now. I'll handle it as separate PR (remind me to file an issue).
5. I sneaked in a minor fix for Globus provider, cause I was too lazy to issue a separate PR (see https://github.com/whole-tale/girder_wholetale/pull/518/commits/b3f1d04b213ffdc5fe57a34916dbf7313987d753). This fixes a case when registered datasets have `identifier` that looks like `doi:doi:<>`.

tl;dr Now any exported Tale with an arbitrary external data (from Zenodo, DataONE, Dataverse or raw HTTP) should be working with:

### How to test?
1. Register datasets from Zenodo, DataONE, Dataverse and register Ligo Tale.
2. Create a Tale that contains a subset of files from above. Mix of subfolder, root dataset folders and individual files would be the best.
3. Export Tale.
4. Verify the bag locally via:
    ```
    $ bdbag --resolve-fetch all .
    $ bdbag --validate full .
    ```
5. Keep the Tale's dataSet somewhere handy. Nuke everything and on a clean install import the bag. Confirm you got the same tale.
